sky/data/storage_utils.py

```diff
@@ -250,6 +269,97 @@ def get_excluded_files(src_dir_path: str) -> List[str]:
     return excluded_paths
 
 
+def is_path_excluded(abs_path: str, base_path: str,
+                     patterns: List[str]) -> bool:
+    """Check if a path matches any exclusion pattern.
+
+    Pattern matching rules (similar to .gitignore):
+    - `*` matches any characters except `/` (within a single path segment)
+    - `**` matches zero or more path segments
+    - `?` matches any single character except `/`
+    - `/` at the beginning anchors the pattern to the root
+    - `/` at the end matches only directories
+
+    Examples:
+    - `*.py` - matches any .py file in any directory
+    - `test/*.py` - matches .py files directly in test/ directory
+    - `test/**/*.py` - matches .py files anywhere under test/ directory
+    - `**/test/**` - matches any path containing /test/ anywhere
+    - `*/test/**` - matches paths like foo/test/bar but not test/bar or
+                    a/b/test/bar
+    - `build/` - matches build directory and everything under it
+    - `./README.md` - matches README.md only at the root
+    - `.env` - matches .env file only at the root
+    - `**/.env` - matches .env files at any depth
+
+    Args:
+        abs_path: Absolute path to check
+        base_path: Base directory path to calculate relative paths from
+        patterns: List of exclusion patterns (from skyignore/gitignore)
+
+    Returns:
+        True if the path should be excluded, False otherwise
+    """
+    rel = os.path.relpath(abs_path, base_path).replace(os.path.sep, '/')
+
+    for pattern in patterns:
+        original_pattern = pattern
+        pattern = pattern.rstrip('/')
+
+        # Handle root-anchored patterns
+        if pattern.startswith('/'):
+            # Remove leading / and treat as root-anchored
+            pattern = pattern[1:]
+            path = rel
+        elif pattern.startswith('./'):
+            # For ./ prefixed patterns, match against ./ + relative path
+            path = './' + rel
+        else:
+            path = rel
+
+        # Handle trailing slash patterns (directory patterns)
+        if original_pattern.endswith('/') and '**' not in pattern:
+            # Pattern like 'build/' should match 'build' and 'build/anything'
+            if path == pattern or path.startswith(pattern + '/'):
+                return True
+            continue
+
+        # Convert glob pattern to regex more simply
+        # Escape special regex chars except * and ?
+        regex_pattern = re.escape(pattern)
+
+        # Handle ** (matches zero or more path segments)
+        # Leading ** should match from start (including empty)
+        regex_pattern = regex_pattern.replace(r'\*\*/', '(?:.*/)?')
```

Comment:
There is weird behavior when this is matched in the middle of the string and the preceding char is not a `/`:
`foo**/bar` -> `foo(?:.*/)?bar` matches `foobar` (no slash in the middle).
Similar comments apply to the following regexes. Is there some reason we can't just keep the last replacement (`\*\*` -> `.*`) and use it to cover all these cases?